sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+The idea is to emit a SPIR-V specialization constant for each aspect that is
+required by a kernel or device function in the device image.  We then introduce
+a new SPIR-V "decoration" that tells the JIT compiler to discard a function if
+a specialization constant is `False`.  The DPC++ runtime will set the values of
+the specialization constants according to the target device, thus the JIT
+compiler discards (and does not compile) any kernels or device functions that
+use features which are not supported on that device.  This avoids errors due to
+speculative compilation of kernels or device functions.
+
+#### Representation in SPIR-V
+
+To illustrate how kernels using optional features are represented in SPIR-V,
+consider a kernel `K` that requires aspects `A1` and `A2`.  The SPIR-V module
+will contain three boolean specialization constants: one representing `A1`, one
+representing `A2`, and one representing the expression `A1 && A2`.  All of
+these can be represented without any extension to SPIR-V.
+
+```
+OpDecorate %11 SpecId 1                       ; External ID for spec const A1
+OpDecorate %12 SpecId 2                       ; External ID for spec const A2
+
+%10 = OpTypeBool
+%11 = OpSpecConstantTrue %10                  ; Represents A1
+%12 = OpSpecConstantTrue %10                  ; Represents A2
+%13 = OpSpecConstantOp %10 LogicalAnd %11 %12 ; Represents A1 && A2
+```
+
+In order to make it easy for the JIT compiler to discard all functions in a
+kernel, each function in the kernel's static call tree (including the function
+representing the kernel's entry point) is decorated with a new extended SPIR-V
+decoration `ConditionalINTEL` whose operand is the `<id>` of the specialization
+constant representing `A1 && A2`.  The semantic of this decoration is that the
+JIT compiler must discard the function unless the value of the specialization
+constant is `True`.  Augmenting the example from above:
+
+```
+OpDecorate %11 SpecId 1                       ; External ID for spec const A1
+OpDecorate %12 SpecId 2                       ; External ID for spec const A2
+OpDecorate %16 ConditionalINTEL %13           ; Says to discard the function
+                                              ; below when (A1 && A2) is False
+%10 = OpTypeBool
+%11 = OpSpecConstantTrue %10                  ; Represents A1
+%12 = OpSpecConstantTrue %10                  ; Represents A2
+%13 = OpSpecConstantOp %10 LogicalAnd %11 %12 ; Represents A1 && A2
+%14 = OpTypeVoid
+%15 = OpTypeFunction %14
+
+%16 = OpFunction %14 None %15                 ; Definition of function that is
+...                                           ; discarded when (A1 && A2) is False
+OpFunctionEnd
+```
+
+See the extension specification of [SpecConditional][6] for a full
+description of this new SPIR-V decoration.
+
+[6]: <extensions/SPIRV/SPV_INTEL_spec_conditional.asciidoc>
+
+#### Representation in LLVM IR
+
+**TODO**: I need some help here on how to represent the `[[sycl::requires()]]`
```

Comment:
I think you are saying that approach (2) is used to represent OpenCL C++ attributes (e.g. `reqd_work_group_size`) in LLVM IR.  Is it safe to assume that we use this approach to represent the SYCL C++ attribute also?  If so, it seems like it would be sensible to use that approach for the new `[[sycl::requires()]]` attribute too.

For example, I think we could represent this:

```
[[sycl::requires(has(A1, A2))]]
void foo() { /* ... */ }
```

as:

```
define void foo() !sycl.required_aspects !1 {
  ret void
}

!1 = !{i32 A1, i32 A2}  ; A1 and A2 are just the integer values of the aspects from "enum class aspect"
```

> The possible downside here is that historically, metadata was introduced to store debug info and AFAIK it could be easily dropped/lost during various optimizations/transformations.

It seems like this must not be a concern if the OpenCL / SYCL C++ attributes are represented this way.  Some of those attributes are required for correctness, so dropping them would cause problems.  In addition, the examples in the [lang ref](https://llvm.org/docs/LangRef.html#function-attributes) like `alignstack(n)` seem to be required for correctness, so I presume they are also not dropped.

Reply:
> Is it safe to assume that we use this approach to represent the SYCL C++ attribute also?

It might be different for each attributes, but ones which exist in OpenCL are represented in the same way.

Reply:
Thanks for all the input.  The latest version of the spec uses metadata to represent the attributes in LLVM IR.